crates/exex/src/manager.rs

```diff
@@ -0,0 +1,446 @@
+                finished_height: None,
+            },
+            event_tx,
+            canon_rx,
+        )
+    }
+
+    /// Reserves a slot in the `PollSender` channel and sends the notification if the slot was
+    /// successfully reserved.
+    ///
+    /// When the notification is sent, it is considered delivered.
+    fn send(
+        &mut self,
+        cx: &mut Context<'_>,
+        (event_id, notification): &(usize, CanonStateNotification),
+    ) -> Poll<Result<(), PollSendError<CanonStateNotification>>> {
+        // check that this notification is above the finished height of the exex if the exex has set
+        // one
+        if let Some(finished_height) = self.finished_height {
+            if finished_height >= notification.tip().number {
+                self.next_notification_id = event_id + 1;
+                return Poll::Ready(Ok(()))
+            }
+        }
+
+        match self.sender.poll_reserve(cx) {
+            Poll::Ready(Ok(())) => (),
+            other => return other,
+        }
+
+        match self.sender.send_item(notification.clone()) {
+            Ok(()) => {
+                self.next_notification_id = event_id + 1;
+                self.metrics.notifications_sent_total.increment(1);
+                Poll::Ready(Ok(()))
+            }
+            Err(err) => Poll::Ready(Err(err)),
+        }
+    }
+}
+
+/// Metrics for the ExEx manager.
+#[derive(Metrics)]
+#[metrics(scope = "exex_manager")]
+pub struct ExExManagerMetrics {
+    /// Max size of the internal state notifications buffer.
+    max_capacity: Counter,
+    /// Current capacity of the internal state notifications buffer.
+    current_capacity: Gauge,
+    /// Current size of the internal state notifications buffer.
+    ///
+    /// Note that this might be slightly bigger than the maximum capacity in some cases.
+    buffer_size: Gauge,
+}
+
+/// The execution extension manager.
+///
+/// The manager is responsible for:
+///
+/// - Receiving relevant events from the rest of the node, and sending these to the execution
+/// extensions
+/// - Backpressure
+/// - Error handling
+/// - Monitoring
+#[derive(Debug)]
+pub struct ExExManager {
+    /// Handles to communicate with the ExEx's.
+    exex_handles: Vec<ExExHandle>,
+
+    /// [`CanonStateNotification`] channel from the [`ExExManagerHandle`]s.
+    handle_rx: UnboundedReceiver<CanonStateNotification>,
+
+    /// The minimum notification ID currently present in the buffer.
+    min_id: usize,
+    /// Monotonically increasing ID for [`CanonStateNotification`]s.
+    next_id: usize,
+    /// Internal buffer of [`CanonStateNotification`]s.
+    buffer: VecDeque<(usize, CanonStateNotification)>,
+    /// Max size of the internal state notifications buffer.
+    max_capacity: usize,
+    /// Current state notifications buffer capacity.
+    ///
+    /// Used to inform the execution stage of possible batch sizes.
+    current_capacity: Arc<AtomicUsize>,
```

Comment:
do we use "size" and "capacity" interchangeably here? If I understand correctly, `max_capacity` is constant and means "maximum size" of internal buffer, while `current_capacity` is "number of elements". Maybe we should clarify it in field names.

Reply:
oh nvm i'm wrong, `current_capacity` is how many you can fit more, all good